File: api/app/routers/lot_image.py
"""Same-origin image proxy for auction CDNs (Copart / IAAI / Encar).

Browsers block or hotlink-break many auction images; locally no-referrer often
works, but production needs /api/lot-image like the Cloudflare Pages Function.
"""

from __future__ import annotations

import re
from urllib.parse import urlparse

import httpx
from fastapi import APIRouter, HTTPException, Query
from fastapi.responses import Response

router = APIRouter(tags=["lot-image"])

_ALLOWED = re.compile(
    r"("
    r"^(?:cs|c-static)\.copart\.(?:com|co\.uk)$|"
    r"(?:^|\.)iaai\.com$|"
    r"(?:^|\.)anvisimages\.com$|"
    r"(?:^|\.)encar\.com$|"
    r"(?:^|\.)bid\.cars$|"
    r"(?:^|\.)cloudfront\.net$|"
    r"(?:^|\.)amazonaws\.com$"
    r")",
    re.I,
)


def _referer_for(host: str, url: str) -> str:
    h = host.lower()
    if "copart.co.uk" in h or "copart.co.uk" in url.lower():
        return "https://www.copart.co.uk/"
    if "copart" in h:
        return "https://www.copart.com/"
    if "iaai" in h or "anvis" in h:
        return "https://www.iaai.com/"
    if "encar" in h:
        return "https://www.encar.com/"
    if "bid.cars" in h:
        return "https://bid.cars/"
    return f"https://{host}/"


@router.get("/api/lot-image")
async def proxy_lot_image(u: str = Query(..., min_length=8)) -> Response:
    target = (u or "").strip()
    try:
        parsed = urlparse(target)
    except Exception as exc:
        raise HTTPException(status_code=400, detail="Bad URL") from exc
    if parsed.scheme not in {"http", "https"} or not parsed.hostname:
        raise HTTPException(status_code=400, detail="Bad URL")
    host = parsed.hostname
    if not _ALLOWED.search(host):
        raise HTTPException(status_code=403, detail="Host not allowed")

    headers = {
        "User-Agent": (
            "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
            "AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36"
        ),
        "Referer": _referer_for(host, target),
        "Accept": "image/avif,image/webp,image/apng,image/*,*/*;q=0.8",
    }
    try:
        async with httpx.AsyncClient(follow_redirects=True, timeout=25.0) as client:
            upstream = await client.get(target, headers=headers)
    except Exception as exc:
        raise HTTPException(status_code=502, detail=f"Upstream fetch failed: {exc}") from exc

    if upstream.status_code == 404:
        raise HTTPException(status_code=404, detail="Not found")
    if upstream.status_code >= 400:
        raise HTTPException(status_code=502, detail=f"Upstream {upstream.status_code}")

    ctype = upstream.headers.get("content-type") or "image/jpeg"
    if "image" not in ctype.lower() and "octet-stream" not in ctype.lower():
        # some CDNs return weird types; still pass through if body looks ok
        ctype = "image/jpeg"

    return Response(
        content=upstream.content,
        status_code=200,
        media_type=ctype,
        headers={
            "Cache-Control": "public, max-age=86400, stale-while-revalidate=604800",
            "Access-Control-Allow-Origin": "*",
        },
    )

File: functions/api/lot-image.js
/**
 * Cloudflare Pages Function — proxies Copart CDN images when needed.
 * Prefer direct CDN URLs in the app; use this via NEXT_PUBLIC_CF_IMAGE_PROXY
 * if a browser environment blocks hotlinking.
 *
 * GET /api/lot-image?u=https://cs.copart.com/...
 */
const ALLOWED_HOST =
  /^cs\.copart\.(com|co\.uk)$|^c-static\.copart\.com$/i;

export async function onRequestGet(context) {
  const reqUrl = new URL(context.request.url);
  const target = reqUrl.searchParams.get("u")?.trim();

  if (!target) {
    return new Response("Bad request", { status: 400 });
  }

  let parsed;
  try {
    parsed = new URL(target);
  } catch {
    return new Response("Bad URL", { status: 400 });
  }

  if (!ALLOWED_HOST.test(parsed.hostname)) {
    return new Response("Host not allowed", { status: 403 });
  }

  const isUk = /\.co\.uk$/i.test(parsed.hostname) || /copart\.co\.uk/i.test(target);
  const upstream = await fetch(target, {
    headers: {
      "User-Agent":
        "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36",
      Referer: isUk ? "https://www.copart.co.uk/" : "https://www.copart.com/",
      Accept: "image/avif,image/webp,image/apng,image/*,*/*;q=0.8",
    },
    cf: { cacheTtl: 86400, cacheEverything: true },
  });

  if (!upstream.ok) {
    return new Response("Upstream error", {
      status: upstream.status === 404 ? 404 : 502,
    });
  }

  const headers = new Headers();
  headers.set("Content-Type", upstream.headers.get("Content-Type") || "image/jpeg");
  headers.set("Cache-Control", "public, max-age=86400, stale-while-revalidate=604800");
  headers.set("Access-Control-Allow-Origin", "*");

  return new Response(upstream.body, { status: 200, headers });
}
